### PR TITLE
Add `parseReadableSize` function

### PR DESCRIPTION
**Pull request by Beetelbrox**
Rework of #64386 implementing the changes requested in #64616
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added SQL functions `parseReadableSize` (along with `OrNull` and `OrZero` variants)

These functions perform the opposite operation of function `formatReadableSize` and `formatReadableDecimalSize`, i.e. the given human-readable byte size, they return the number of bytes. Example: `SELECT formatReadableSize('3.0 MiB')` returns `3145728`.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All NOT Required Checks
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: Integration Tests
- [ ]  Exclude: Stateless tests
- [ ]  Exclude: Stateful tests
- [ ]  Exclude: Performance tests
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with Aarch64
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
---
- [ ]  Do not test
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit f5234af77070d061e2132455a00eecfc63b7e3bc with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64742/f5234af77070d061e2132455a00eecfc63b7e3bc/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: docs/en/sql-reference/functions/other-functions.md [hunk 1]
```diff
@@ -890,6 +894,122 @@ SELECT
 └────────────────────┴────────────────────────────────────────────────┘
 ```
 
+## parseReadableSize
+
+Given a string containing a byte size and `B`, `KiB`, `KB`, `MiB`, `MB`, etc. as a unit (i.e. [ISO/IEC 80000-13](https://en.wikipedia.org/wiki/ISO/IEC_80000) or decimal byte unit), this function returns the corresponding number of bytes.  
+If the function is unable to parse the input value, it throws an exception.
+
+The inverse operations of this function are [formatReadableSize](#formatReadableSize) & [formatReadableDecimalSize](#formatReadableDecimalSize).
```
Comment (rschu1ze) [inline docs/en/sql-reference/functions/other-functions.md:902]: `&` --> `and` (same in l. 940/978)

File: src/Functions/parseReadableSize.cpp [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+enum class ErrorHandling : uint8_t
+{
+    Exception,
+    Zero,
+    Null
+};
+
+using ScaleFactors = std::unordered_map<std::string_view, size_t>;
+
+/** parseReadble*Size - Returns the number of bytes corresponding to a given readable binary or decimal size.
```
Comment (rschu1ze) [inline src/Functions/parseReadableSize.cpp:41]: `parseReadble*Size` --> `parseReadableSize*`

File: src/Functions/parseReadableSize.cpp [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args
+        {
+            {"readable_size", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), nullptr, "String"},
+        };
+        validateFunctionArgumentTypes(*this, arguments, args);
+        DataTypePtr return_type = std::make_shared<DataTypeUInt64>();
+        return (error_handling == ErrorHandling::Null) ? std::make_shared<DataTypeNullable>(return_type) : return_type;
```
Comment (rschu1ze) [inline src/Functions/parseReadableSize.cpp:73]: Could use `if constexpr` here (though it doesnt' make a performance difference). Also, for slightly more complex code it is nicer to not use a ternary if. So let's do:

```cpp
if constexpr (error_handling == ErrorHandling::Null)
    [...]
else
    [...]
```

File: docs/en/sql-reference/functions/other-functions.md (hunk 1 above)
Comment (Beetelbrox) [reply docs/en/sql-reference/functions/other-functions.md]: applied

File: src/Functions/parseReadableSize.cpp (hunk 3 above)
Comment (Beetelbrox) [reply src/Functions/parseReadableSize.cpp]: applied

File: src/Functions/parseReadableSize.cpp (hunk 2 above)
Comment (Beetelbrox) [reply src/Functions/parseReadableSize.cpp]: applied

**Review by rschu1ze — APPROVED** [review]

**Comment by rschu1ze** [comment]
None of the failures in CI is related, a sync is green. Lgtm, forcing MergeableCheck green.

**Pull request merged**